mono/metadata/class.c

```diff
@@ -515,6 +521,18 @@ mono_type_get_name_recurse (MonoType *type, GString *str, gboolean is_recursed,
 			MonoTypeNameFormat nested_format;
 			int i;
 
+			// https://bugzilla.xamarin.com/show_bug.cgi?id=57938
+			// Generic instances containing generic parameters from themselves
```

Comment:
Edit: Actually now that I read the .NET docs I see my earlier (deleted) comments were wrong.  So now I just have a single nit to pick: "generic parameters from themselves" is wrong: https://docs.microsoft.com/en-us/dotnet/api/system.type.fullname?view=netframework-4.7.1#System_Type_FullName

> The fully qualified name of the type, including its namespace but not its assembly; or `null` if the current instance represents a generic type parameter, an array type, pointer type, or byref type based on a type parameter, or a generic type that is not a generic type definition but contains unresolved type parameters.

This doesn't say that the type parameters have to be from the same gtd.   it seems like if the generic instance is open the `FullName` should be null.

so something like:
```c
  if (!mono_class_is_gtd (klass) && mono_class_is_open_constructed_type (klass)) {
    g_string_truncate (str, 0);
    return;
  }
```

And I'd test it on something like:

```csharp
  class Foo<A,B> { }
  class Bar<X> : Foo <Int, List <X>> { }
  // expect typeof (Bar`1).BaseType.FullName to be NULL
```